tests/test_compiler.py

```diff
@@ -206,3 +206,23 @@ def circuit(x: float):
         mlir_str = str(circuit.mlir)
         result_header = "func.func private @circuit(%arg0: tensor<f64>) -> tensor<f64>"
         assert result_header in mlir_str
+
+    def test_control(self):
+        """Test that control works with qjit."""
+        dev = qml.device("lightning.qubit", wires=2)
+
+        @qml.qjit
+        @qml.qnode(dev)
+        def workflow(theta, w, cw):
+            qml.Hadamard(wires=[0])
+            qml.Hadamard(wires=[1])
+
+            def func(arg):
+                qml.RX(theta, wires=arg)
+
+            qml.ctrl(func, control=[cw])(w)
```

Comment:
The reason we allow this is because it's impossible for catalyst to know whether two values are the same or not until the program is executed. Consider the following function:

```py
@qjit
@qml.qnode
def circuit(a, b):
    qml.Hadamard(0)
    qml.CNOT(wires=[a, b])
    return qml.probs()
```

`a` and `b` are not until the user provides, so we cannot perform any checks. At runtime the device may throw an error if the program is invalid (e.g. by aliasing two wire values to the same logical qubit index).
Pennylane would face the issue with `jax.jit` if it allowed JAX tracers as wires.

Reply:
To answer your second question, there is in fact a variety of new features / programming paradigms that Catalyst unlocks that are currently not supported in PennyLane, simply due to the restrictions of PL's architecture . One of the core goals of Catalyst was to support dynamic quantum programming paradigms. Examples of this are shown in our documentation, such as adaptive circuits and repeat-until-success algorithms.

Reply:
hmm, fair enough. my concern was that the device did _not_ throw an error, presumably because the wires are jitted and can't be validated. I just checked both `qml.cond` and `qml.CNOT` like this to see what would happen, and they both fail without qjit, but pass with it.

is this something we should document? probably part of the larger problem of wires being hashable/jittable in PL

Reply:
I just double checked and the following does indeed run without throwing errors:
```py
@qjit
@qml.qnode(qml.device("lightning.qubit", wires=2))
def f(a: int, b: int):
    qml.Hadamard(0)
    qml.CNOT(wires=[a, b])
    return qml.probs()
f(0, 0)
```
I would file this as a bug, since either the catalyst runtime (or the lighting devices) should raise an error on that CNOT (probably both).

Reply:
there is some discussion in #4481 on this question already (allowing for jittable wires), and I don't foresee changes in PL core about this, but I'm happy to re-open that discussion. if you think this should be fixed in catalyst (or not fixed at all?) that works too

Reply:
Yes, I've taken note of this bug in the Catalyst runtime and it will hopefully be fixed by next release on our end :)